### ISSUE TITLE
Areas of discomfort related to Ceylon, noticed during the writing of ceylon-dddsample 

### ISSUE THREAD

**Issue by sgalles** [sgalles/ceylon-dddsample#1]
@gavinking this is my summary

First I must say that overall I really like what Ceylon brings to a JEE application. There are areas of discomfort, but I must say that it is super pleasant to see the LOC-melt from Java to Ceylon.
And also, the JPA entities looks very good in Ceylon.

That being said, the problems now :
## Ceylon `WarInitializer` and JPA

Ceylon `WarInitializer` leverages a servlet initialization to initialize Ceylon in a war. Problem is, the servlet initialization does not kick in so enough during the container startup in order to handle all cases.

For instance here a singleton EJB is used to bootstrap the model of the application in database https://github.com/sgalles/ceylon-dddsample/blob/master/source/dddsample/cargotracker/application/util/SampleDataGenerator.ceylon but this happens before the servlet init. 

A workaround was added in `ceylon-dddsample` for this problem :

The `WarInitializer` servlet is deactivated, and is replaced by almost the same initialization in a portable CDI `Extension` https://github.com/sgalles/ceylon-dddsample/blob/master/source/dddsample/CeylonInit.java

Problem is, this scheme is not generic enough to be proposed as an alternative to the `WarInitializer` servlet, indeed :
- it has a dependency on `org.jboss.vfs.VirtualFile` because that's the only way that I found to emulate the behavior of  `WarInitializer` without a `ServletContext` available 
- Even with the help of `org.jboss.vfs.VirtualFile` I had to find a workaround for https://issues.jboss.org/browse/JBVFS-147 and I leverage a side effect of the implementation of `org.jboss.vfs.VirtualFile`. It probably means that `CeylonInit` is not even Wildfly-portable...

FTR Here https://groups.google.com/forum/#!msg/ceylon-users/Lbcnm3ac9PA/hJgZxcqZAwAJ @jvasileff provides a general solution to this initialization problem.
## Ceylon collections not usable in JPA entities

Is suspect that not much can be done about this, but basically all collections in JPA entities must be plain Java collections, else JPA complains (IIRC it did not work when I tried to use the collection wrappers of the SDK, but I may have to check this out again).
Update : as suggested by @gavinking, Hibernate `Tuplizers` should help.
## `javax.persistence.Converter` does not work in `autoApply`  mode

It is minor, but I had to manually apply the converter declared here https://github.com/sgalles/ceylon-dddsample/blob/master/source/dddsample/cargotracker/domain/infrastructure/persistence/jpa/CeylonEnumValuesConverter.ceylon because when I tried to set `autoApply=true` I received strange classloader errors IIRC. Well not a big deal though, the manual mode works like a charm https://github.com/sgalles/ceylon-dddsample/blob/master/source/dddsample/cargotracker/domain/model/cargo/Delivery.ceylon#L50
### Ceylon default constructors in all JPA entities

Was fixed by : 
https://github.com/ceylon/ceylon-spec/issues/1405
### Ceylon language types not Serializable (for passivating scope)

Was fixed by :
https://github.com/ceylon/ceylon-compiler/issues/2329
### JSF conversion error caused by automatic conversion `ceylon.language.String` -> `java.util.String`

Basically, because of the automatic conversion I got this error 
http://stackoverflow.com/questions/9069379/validation-error-value-is-not-valid
Solved by using `String?` instead of `String` it deactivates the conversion.
https://groups.google.com/forum/#!msg/ceylon-users/5n6Oidxa-94/ZaQesjNDBgAJ

**Comment by luolong** [comment]
> Ceylon default constructors in all JPA entities

Not only JPA entities. Also in JAX-RS resources and generally in all other places where we want to add class or constructor level annotations...

**Comment by luolong** [comment]
> Ceylon collections not usable in JPA entities

Also, they break Jackson (and possibly other JSON/XML serializer libraries)

**Comment by sgalles** [comment]
> Also, they break Jackson

Ah yes indeed @luolong 

@gavinking I've updated the list to add _Ceylon language types not Serializable_ because JSF mandates that ViewScope or SessionScope backing beans be serializable, but Ceylon objects are not java-serializable (String for instance) so basically Ceylon is not a good fit for a JSF backing beans (at least for the View and Session scopes)

**Comment by gavinking** [comment]
@sgalles did you try using a [CDI Portable Extension](http://docs.jboss.org/cdi/api/1.0/javax/enterprise/inject/spi/package-summary.html) for bootstrapping Ceylon?

**Comment by gavinking** [comment]
@sgalles to make Ceylon work really well with Hibernate, we surely need to write a custom [Tuplizer](https://docs.jboss.org/hibernate/orm/3.5/api/org/hibernate/tuple/Tuplizer.html) for Ceylon classes.

**Comment by gavinking** [comment]
So I made some basic types serializable: `Float`, `Integer`, `String`, `Character`, `Boolean`, `Byte`, `Tuple`, and `Array`.

But that doesn't give us any really useful container classes, not even `ArraySequence`, `ArrayList`, `HashMap`, nor `HashSet`.

Sure, you could use collections from `java.util` to back your JavaBeans, and wrap them in the getters, but that's a pain.

**Comment by sgalles** [comment]
@gavinking 

> did you try using a CDI Portable Extension for bootstrapping Ceylon?

Ah ! no, but I'm definitly going to try this out

> to make Ceylon work really well with Hibernate, we surely need to write a custom Tuplizer for Ceylon classes.

yes, indeed, for `ceylon.language.String` I could leverage `javax.persistence.Converter` but there's nothing in JPA to convert new collections types. And I happily admit that I didn't know about Tuplizer. Interesting. 

> So I made some basic types serializable

Yahoo! neat.

> Sure, you could use collections from java.util to back your JavaBeans, and wrap them in the getters, but that's a pain

Well anyway I already did something like this to make Ceylon collections work with JPA
https://github.com/sgalles/ceylon-dddsample/blob/master/source/dddsample/cargotracker/domain/model/cargo/Itinerary.ceylon#L56

**Comment by sgalles** [comment]
FTR I updated the description of the issue with the entry
`JSF conversion error caused by automatic conversion ceylon.language.String -> java.util.String`

**Comment by sgalles** [comment]
FTR Updated the description of the issue to reflect the change from a `PersistenceProvider` to a CDI portable `Extension` (thanks @gavinking for the suggestion)

**Comment by gavinking** [comment]
@sgalles thanks to @tombentley, (virtually) every Ceylon class now implements `Serializable`, and has a default constructor. HTH!

**Comment by sgalles** [comment]
Yeah ! It definitly helps.
Massive code simplification !
https://github.com/sgalles/ceylon-dddsample/commit/4c8259fd173829863e3745ab6f8742f8a3eebec4
https://github.com/sgalles/ceylon-dddsample/commit/a86551b44def051163f6cbf6c020542644bd4da7
Thanks @tombentley and @gavinking

**Comment by sgalles** [comment]
Description updated : added link to ceylon/ceylon#2427 (Java meta-annotation)

**Comment by gavinking** [comment]
> Is suspect that not much can be done about this, but basically all collections in JPA entities must be plain Java collections, else JPA complains (IIRC it did not work when I tried to use the collection wrappers of the SDK, but I may have to check this out again).

This is fixed with the new "EE mode" in Ceylon 1.3.1.

**Comment by gavinking** [comment]
> Ceylon `WarInitializer` leverages a servlet initialization to initialize Ceylon in a war. Problem is, the servlet initialization does not kick in so enough during the container startup in order to handle all cases.

I might be wrong, but I think that this should in principle be solved by `ceylon war --static-metamodel`, right, @FroMage?

**Comment by FroMage** [comment]
Yes.

**Comment by sgalles** [comment]
@FroMage @gavinking I removed my CDI Extension and tried to use `--static-metamodel` instead (was on Wildfly 9) but it seems that the Ceylon metamodel initialization does not kick in soon enough during the Ceylon initialization (and some of my Ceylon classes that rely on the metamodel breaks during the initialization of my app). How does `--static-metamodel` work ? what trigger does it use to init the metamodel ? Is there a way to force it to init sooner ?

**Comment by gavinking** [comment]
> and some of my Ceylon classes that rely on the metamodel breaks during the initialization of my app

Could you tell me more about those classes and what they do?

> How does `--static-metamodel` work ? what trigger does it use to init the metamodel ? Is there a way to force it to init sooner ?

@FroMage ?

**Comment by FroMage** [comment]
> How does --static-metamodel work ? what trigger does it use to init the metamodel ? Is there a way to force it to init sooner ?

It should be loaded as soon as the metamodel is used. Can you point me to code to reproduce the issue?

**Comment by sgalles** [comment]
Yes @FroMage see my comment here https://github.com/sgalles/ceylon-dddsample/issues/2#issuecomment-263088971

**Comment by sgalles** [comment]
So, now that `master` works with Ceylon 1.3.1 I'm going to leverage the EE mode to clean up the few remaining minor issues

**Comment by gavinking** [comment]
Great, looking forward to seeing if it helps.

**Comment by sgalles** [comment]
So, enventually I could :
* remove the custom init that used a CDI Extension (that's great !)
* fix this problem with CDI qualifiers (not related too EE mode though) : https://github.com/sgalles/ceylon-dddsample/commit/247dfa3ca943422fb9229cc8a5413f17cbb8009c
* remove all the `default` in CDI beans : https://github.com/sgalles/ceylon-dddsample/commit/31d121f6e675fcade133d8a7357bc60b3b05257f
* remove the use of JList for JPA : here https://github.com/sgalles/ceylon-dddsample/commit/728c95d0f7bea4d0bce8d864d6fe5ca472f7b208 and here https://github.com/sgalles/ceylon-dddsample/commit/39273a04ce1ec18acab9f7073c3239fc4ec1d099
* remove a bunch of `\i` (still not related to EE mode) https://github.com/sgalles/ceylon-dddsample/commit/6397f36973916e505835f17170ed2c3e79a4edac

So, now, for me, the use of Ceylon with JAVA EE is much more intuitive. 

Maybe there still are a few remaining things left interop-wise :
* I think I'll try to remove the JString from this `javax.ws.rs` class https://github.com/sgalles/ceylon-dddsample/blob/master/source/dddsample/pathfinder/api/GraphTraversalService.ceylon#L49 and use Ceylon Strings
* for JSF I had to use a lost of JList as JSF can't use FIELD access and it can't leverage the EE mode AFAICT. But I guess there isn't much that can be done here (but I can't say that's a problem for me, I don't really use JSF these days...)

**Comment by gavinking** [comment]
@sgalles Sounds great!

I'm trying to find a moment to review the example app.
